app/backend/core/authentication.py

```diff
@@ -0,0 +1,202 @@
+            raise AuthError(
+                {"code": "invalid_header", "description": "Authorization header must start with Bearer"}, 401
+            )
+        elif len(parts) == 1:
+            raise AuthError({"code": "invalid_header", "description": "Token not found"}, 401)
+        elif len(parts) > 2:
+            raise AuthError({"code": "invalid_header", "description": "Authorization header must be Bearer token"}, 401)
+
+        token = parts[1]
+        return token
+
+    @staticmethod
+    def build_security_filters(overrides: dict[str, Any], auth_claims: dict[str, Any]):
+        # Build different permutations of the oid or groups security filter using OData filters
+        # https://learn.microsoft.com/azure/search/search-security-trimming-for-azure-search
+        # https://learn.microsoft.com/azure/search/search-query-odata-filter
+        use_oid_security_filter = overrides.get("use_oid_security_filter")
+        use_groups_security_filter = overrides.get("use_groups_security_filter")
+
+        oid_security_filter = (
+            "oids/any(g:search.in(g, '{}'))".format(auth_claims.get("oid") or "") if use_oid_security_filter else None
+        )
+        groups_security_filter = (
+            "groups/any(g:search.in(g, '{}'))".format(", ".join(auth_claims.get("groups") or []))
+            if use_groups_security_filter
+            else None
+        )
+
+        # If only one security filter is specified, return that filter
+        # If both security filters are specified, combine them with "or" so only 1 security filter needs to pass
+        # If no security filters are specified, don't return any filter
+        if oid_security_filter and not groups_security_filter:
+            return oid_security_filter
+        elif groups_security_filter and not oid_security_filter:
+            return groups_security_filter
+        elif oid_security_filter and groups_security_filter:
+            return f"({oid_security_filter} or {groups_security_filter})"
+        else:
+            return None
+
+    @staticmethod
+    async def list_groups(graph_resource_access_token: str) -> [str]:
+        headers = {"Authorization": "Bearer " + graph_resource_access_token["access_token"]}
+        resp = urllib3.request(
+            "GET",
+            "https://graph.microsoft.com/v1.0/me/memberOf?$select=id",
+            headers=headers,
+            timeout=urllib3.Timeout(connect=10, read=10),
+        )
+        groups = []
+        while resp.status == 200:
+            value = resp.json()["value"]
+            for group in value:
+                groups.append(group["id"])
+            nextLink = resp.json().get("@odata.nextLink")
+            if nextLink:
+                resp = urllib3.request("GET", nextLink, headers=headers, timeout=urllib3.Timeout(connect=10, read=10))
+            else:
+                break
+
+        return groups
+
+    async def get_auth_claims_if_enabled(self) -> dict[str:Any]:
```

Comment:
Just curious, is this code similar to https://github.com/snok/django-auth-adfs/blob/cc700730a4130365dbd6094612f1cf57ad2a3d6f/django_auth_adfs/backend.py#L196 ?
Might be useful to compare your approach if so, I know you mentioned there was some trickiness to group claim checking.

Reply:
my approach is pretty much the same. the only difference is they are using the transitive member of, which I can switch to

Reply:
i'll keep it member of for now

Reply:
made the change